### ISSUE TITLE
Nightly regression on x86_64_unknown_linux_gnu

### ISSUE THREAD

**Issue by Aceeri** [rust-lang/rust#43126]
Travis is failing for me on the newer nightlies, but not previous ones and not on stable/beta. I'm willing to guess it has something to do with procedural macros, since other PRs build correctly on this nightly.

Travis builds for comparison:
Success: [rustc 1.19.0-nightly (10d7cb44c 2017-06-18)](https://travis-ci.org/slide-rs/specs/builds/251552764)
Failure: [rustc 1.19.0-nightly (04145943a 2017-06-19)](https://travis-ci.org/slide-rs/specs/builds/251552066)

Related PR: https://github.com/slide-rs/specs/pull/201

Exits with code 101 with no error messages, just a failure. Can't reproduce on Windows, so I am kind of limited in trying to debug this.

**Comment by sfackler** [comment]
Seems to work for me (at least locally) on `rustc 1.20.0-nightly (9b85e1cfa 2017-07-07)` on Linux.

**Comment by TimNN** [comment]
Ok, I think there's three things going on:

1. `rustup` is annoying and turns a segmentation fault produced by `rustc` into an exit code 1
2. The failing `rustc` invocation uses `serde_derive`, that is it runs code previously produced by the compiler.
3. IIRC the `2017-06-20` nightly falls into the range of ["duplicate drop"](https://github.com/rust-lang/rust/issues/42903) nightlies and thus produces bad code.

**Comment by Mark-Simulacrum** [comment]
It'd be good to verify if this is fixed by rerunning on a more recent nightly. Marking as a regression for now though...

**Comment by Aceeri** [comment]
@Mark-Simulacrum I just re-ran the travis build and now it seems both `nightly` and `beta` have this issue now:
https://travis-ci.org/slide-rs/specs/builds/255438162?utm_source=github_status&utm_medium=notification

**Comment by Mark-Simulacrum** [comment]
I can't seem to reproduce locally on macOS either...

**Comment by alexcrichton** [comment]
@Aceeri would it be possible to make a standalone reproduction of this? Right now I'm not even sure how to get access to the commit on my local machine...

**Comment by Aceeri** [comment]
@alexcrichton I'll see if I can try a couple of things to reproduce it if I can find some time.

**Comment by Marwes** [comment]
Might have the same bug in combine https://travis-ci.org/Marwes/combine/jobs/261073809. Compiling on windows with nightly works fine as well.

**Comment by Marwes** [comment]
Reproduction: [gist](https://gist.github.com/e8b7ea953b4877a02d06b0082f85a5f4) [playground](https://play.rust-lang.org/?gist=e8b7ea953b4877a02d06b0082f85a5f4&version=nightly)

Seems like for some reason x86_64_unknown_linux_gnu segfaults(?) when trying to compile the trait implementation of an extremely nested type. Each tuple implementation in the gist expands a type as `Or<A, Or<B, Or<C, ...>>>` so it compiles if I implement it only for tuples of lesser arity or if I break up the implementation to have a be wider instead of so deep.

**Comment by Aceeri** [comment]
That might actually be the problem, I have a macro in mine that recurses ~128 times.

**Comment by alexcrichton** [comment]
@Marwes I believe your example here may be the same as https://github.com/rust-lang/rust/issues/43613, where the segfault is just manifesting as OOM. I ran it locally and rustc ran way with half my memory before I killed it.

In that case I wonder if we can tag https://github.com/rust-lang/rust/issues/43613 as a regression and close this?

**Comment by nikomatsakis** [comment]
@TimNN @Mark-Simulacrum -- any chance either of you want to try and [bisect the more narrow example that @Marwes gave here](https://github.com/rust-lang/rust/issues/43126#issuecomment-320436582)?

**Comment by nikomatsakis** [comment]
triage: P-high -- seems like a valid regression we need to investigate.

**Comment by TimNN** [comment]
According to `rust-bisect`, this regressed from a successful compilation to an OOM (when given 500MB ram) with https://github.com/rust-lang/rust/pull/43546.

**Comment by TimNN** [comment]
I attempted to bisect the commit from https://github.com/rust-lang/rust/issues/43126#issuecomment-316532163 (again with 500MB of ram, I couldn't repro on the latest nightly with full memory). The results were varied (sometimes a segmentation fault and sometimes an illegal instructions, so I'm not sure how helpful or correct the result will be).

`rust-bisect` (consistently) points to #40847, which however falls outside the failed / success range in the OP (the failing commit from the OP was no longer available).

---

Event though I couldn't repro on the latest nightly with 1GB ram, I ran `rust-bisect` again with 1GB ram and got #39409, which does fall into the range in the OP.

**Comment by TimNN** [comment]
@alexcrichton: I did the bisection from my last comment again, this time with an 8 GB ram limit (which is much larger than the 4GB the OP's failing builds had on travis (if they ran on the container infrastructure)).

So I currently believe that some or all of #43546, #40847 and #39409 (and/or maybe others) caused an increase in ram requirements (which may have decreased again in the latest nightly).

**Comment by alexcrichton** [comment]
@TimNN are you on Linux? We could try to test this out with massif and valgrind. If you've got the exact compilers and nightlies I can also try to run locally.

In general though you can just do:

```rust
valgrind --tool=massif --trace-children=yes rustc ...
ms_print massif.* | less
```

And then there should be a "pretty graph" at the top of the output

**Comment by nikomatsakis** [comment]
@pnkfelix will investigate whether #39409 is indeed at fault, as it it seems the most likely culprit from those 3.

**Comment by alexcrichton** [comment]
1.20 is now stable

**Comment by nikomatsakis** [comment]
Update: @pnkfelix mentioned to me that he was going to attempt to land a `-Z` flag that would let you disable `EndRegion` emission, so that we can test the impact more easily. I don't think this has happened yet.

**Comment by nikomatsakis** [comment]
https://github.com/rust-lang/rust/pull/44219 is the PR.

**Comment by pnkfelix** [comment]
@alexcrichton BTW I had a really hard time getting useful data out of massif (it was claiming the peak memory usage was something like 12kb), until last night @eddyb told me that valgrind/massif do not behave well with jemalloc.

So after I hacked the compiler to use the system allocator (via the diff below), I got *much* more believable results from valgrind/massif.
  * (This leads me to wonder if we should provide an easier way for the user to opt into using the system allocator when running `rustc`... I don't know if making it a dynamic choice when running `rustc` will totally defeat the gains of providing jemalloc as an option in the first place, but I figure I should at least mention the idea...)

```
diff --git a/src/rustc/rustc.rs b/src/rustc/rustc.rs
index bfd01146d2..1c3d8b1da2 100644
--- a/src/rustc/rustc.rs
+++ b/src/rustc/rustc.rs
@@ -9,7 +9,9 @@
 // except according to those terms.

 #![feature(rustc_private)]
+#![feature(alloc_system)]

+extern crate alloc_system;
 extern crate rustc_driver;

 fn main() { rustc_driver::main() }
```

**Comment by nikomatsakis** [comment]
Can we confirm if this is fixed (for now) by #44249?

**Comment by nikomatsakis** [comment]
Going to assume this was resolved by #44249. Closing. Please let us know if you are still seeing problems, @Aceeri!
